modules/memfault-firmware-sdk/memfault_gnss_metrics.c

```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <modem/location.h>
+#include "memfault/metrics/metrics.h"
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_DECLARE(memfault_ncs_metrics, CONFIG_MEMFAULT_NCS_LOG_LEVEL);
+
+static bool session_in_progress = false;
+static location_event_handler_t real_location_event_handler;
+
+static void memfault_location_event_handler(const struct location_event_data *event_data);
+static void memfault_gnss_metrics_start_fix_session(void);
+static void memfault_gnss_metrics_stop_fix_session(const struct location_event_data *event_data);
+
+// To maintain separation of metrics collection from NCS library code, we need a way to
+// know when a fix request is made, and when the result of the request is ready (fix or error).
+// There is a location event handler that indicates a location request result, but only one,
+// which should be reserved for application code. There is also no callback mechanism to
+// be notified when a fix request is made. Therefore, we wrap the functions of interest:
+//        - location_init() -- registers an event callback. Save the application
+//              callback and register a wrapper callback.
+//        - location_request() -- initiates a request. Intercept it's result, and
+//              start a session if the request is successful.
+int __real_location_init(location_event_handler_t event_handler);
+int __wrap_location_init(location_event_handler_t event_handler)
+{
+	real_location_event_handler = event_handler;
+	return __real_location_init(memfault_location_event_handler);
+}
+
+// Ensure the substituted function signature matches the original function
+_Static_assert(__builtin_types_compatible_p(__typeof__(&location_init),
+					    __typeof__(&__wrap_location_init)) &&
+		       __builtin_types_compatible_p(__typeof__(&location_init),
+						    __typeof__(&__real_location_init)),
+	       "Error: Wrapped function does not match original location_init function signature");
+
+// Also wrap the location request to know when to start a session
+int __real_location_request(const struct location_config *config);
+int __wrap_location_request(const struct location_config *config)
+{
+	int result = __real_location_request(config);
```

Comment:
How does this work if there is another library trying to do similar tricks with linker options?

Reply:
Typically, the first `__wrap_*` matching symbol will be used. However, since Zephyr forces `ld` to look at everything when resolving symbols instead of choosing the first match as goes, if another library tries to wrap the same function, we'll get a compiler error like:

```
wrap2.c:(.text+0x0): multiple definition of `__wrap_foo'; wrap1.o:wrap1.c:(.text+0x0): first defined here
```

Reply:
Ok, linking error is good so things won't happen behind the scenes without nobody knowing what method instance is actually used.

Reply:
Correct. I think wrapping is a huge hack, so if we can avoid it, that would be ideal honestly. Sounds like we can if we can figure out some callback scheme